nat-lab/bin/netcat.py

```diff
@@ -0,0 +1,209 @@
+import argparse
+import errno
+import selectors
+import socket
+import sys
+import time
+
+UDP_SCAN_COUNT: int = 5
+RECV_SIZE = 4096
+
+
+class NetCat:
+    """A netcat clone written in Python"""
+
+    def __init__(self, args):
+        self.args: argparse.Namespace = args
+        self.verbose: bool = args.v
+        self.udp: bool = self.args.u
+        self.sock_type: str = "udp" if self.udp else "tcp"
+        self.listen: bool = self.args.l
+        self.ipv6: bool = self.args.ipv6
+        self.sock: socket.socket = self._create_socket()
+        self.client_addr: str | None = None
+        self.selector = selectors.DefaultSelector()
+        self.should_close = False
+
+    def _vprint(self, *args, **kwargs):
+        """print an event to stderr in verbose mode"""
+        if self.verbose:
+            print(*args, **kwargs, file=sys.stderr)
+
+    def _create_socket(self) -> socket.socket:
+        """Helper to create socket"""
+        family = socket.AF_INET6 if self.ipv6 else socket.AF_INET
+        sock_type = socket.SOCK_DGRAM if self.udp else socket.SOCK_STREAM
+        try:
+            s = socket.socket(family, sock_type)
+            return s
+        except OSError as e:
+            print(f"error creating socket: {e}", file=sys.stderr)
+            sys.exit(1)
+
+    def _connect(self):
+        """Connect to a remote listener"""
+        hostname = self.args.hostname
+        port = self.args.port
+
+        try:
+            # Bind to a local port or source address if specified
+            if self.args.p or self.args.s:
+                local_address: str = self.args.s if self.args.s else ""
+                local_port: int = self.args.p if self.args.p else 0
+
+                self.sock.bind((local_address, local_port))
+
+            self.sock.connect((hostname, port))
+            if self.udp:
+                self._udptest()
+            self._vprint(
+                f"Connection to {hostname} {port} port [{self.sock_type}/*] succeeded!"
+            )
+        except OSError as e:
+            if e.errno == errno.EBADF:
+                # connection closed
+                sys.exit(0)
+            print(
+                f"nc: connect to {hostname} port {port} ({self.sock_type}) failed: {e.strerror}",
+                file=sys.stderr,
+            )
+            sys.exit(1)
+
+    def _listen(self):
+        """Listen for incoming connections"""
+        # Set SO_REUSEPORT option
+        try:
+            self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
+        except AttributeError:
+            # SO_REUSEPORT may not be available on all systems, fall back to SO_REUSEADDR
+            self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+
+        port = self.args.port
+        bind_addr = "::" if self.ipv6 else "0.0.0.0"
```

Comment:
I think you are missing `source_ip` here. If it is set for server mode -> we should bind server-side sockets to some specific source IP as well.

Reply:
Thanks, good catch!